runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -314,6 +321,32 @@ function LanguageTree:_get_injections()
         end
       end
 
+      for _, pred in pairs(preds or {}) do
+        -- Shifts the column range of the injected node and uses
+        -- the new range as the injection point. This should be used
+        -- when the offset is known (stripping quotes, for example).
+        -- Example: (#offset! @_node 0 1 0 -1)
+        if pred[1] == "offset!" then
+          local offset_node = match[pred[2]]
+          local range = {offset_node:range()}
+          local start_row_offset = pred[3] or 0
```

Comment:
Wouldn't be better if we could also use this predicate in other queries (not only injections)? This directive could also just replace the node by a range table (or save the processed result some where else in `info`) and by that be useful for other applications that need to specify an offset on their query.

Wouldn't it make sense to also allow strings as offset or more general Lua expressions? E.g. `(offset! 0 4 0 0)` could also be written as `(offset! 0 ">>> " 0 0)` or `(offset! 0 "#'>>> '" 0 0)`? But probably just numbers is the fastest.

Would be nice if pred could `gmatch` on a node range to get all relevant subranges e.g. to find all `>>> (.*)$` in Python docstrings? I assume most users won't want to implement their on tree-sitter parser for tiny things like this. Also this is a concept that should not be specific to injections only.

Reply:
It could be made more general but this use case is very specific to injections. we don't have anywhere to put the resulting range that this generates currently... I'm not opposed to storing the results in `info` but it needs something more generic.

This predicate is meant to be a fixed offset. Only really used when the offset is known by the user, and is meant to be as fast as possible. Any pattern matching should be done in a separate predicate to keep this one fast and not overloaded.

Also, stripping content from the lines of the captured node isn't supported in this PR. That will come at a later time. This covers about 80% of use cases when it comes to stripping unneeded characters from language injection.

Reply:
I'm just saying that we might need more such things for injection needs. Would be nicer if we could have something that is available everywhere and not just hidden magic for injections (users might try to use #offset! also for highlights or something else)

`#offset!` would be also nice for textobjects and those situations in highlights where the users want to trim a leading character.

The more general version with string is bullshit. I admit it! :smile:

Reply:
But this PR is a good incremental improvement. We can replace with something more general later.

Reply:
I agree, this would be better abstracted into something more reusable. I'm approaching this in two different phases.

- Basic injection offsets for removing leading/trailing quotes or lines. "`fennel(-?> tbl (. :key))`"
- Second is doing "fenced injection" where we strip characters per line of the node range

```
/**
 * @example
 * for (const item of items) {
 *   console.log(item);
 * }
 */
```
Sending only the javascript above to the parser and stripping the `*` before each line (this one is going to be tricky).

I'll look into an abstraction to make this more general use.